src/metadata.jl

```diff
@@ -86,75 +86,171 @@ function typestr(s::AbstractString, filterlist=nothing)
     end
 end
 
+
 """Metadata configuration of the stored array
 
 Each array requires essential configuration metadata to be stored, enabling correct
 interpretation of the stored data. This metadata is encoded using JSON and stored as the
-value of the “.zarray” key within an array store.
+value of the ".zarray" key within an array store.
+
+# Type Parameters
+* T - element type of the array
+* N - dimensionality of the array
+* C - compressor
+* F - filters
+
+# See Also
 
 https://zarr.readthedocs.io/en/stable/spec/v2.html#metadata
 """
-struct Metadata{T, N, C, F}
+abstract type AbstractMetadata{T,N,C,F} end
+Base.ndims(::AbstractMetadata{<:Any,N}) where N = N
+
+
+"""Metadata for Zarr version 2 arrays"""
+struct MetadataV2{T,N,C,F} <: AbstractMetadata{T,N,C,F}
     zarr_format::Int
+    node_type::String
     shape::Base.RefValue{NTuple{N, Int}}
     chunks::NTuple{N, Int}
     dtype::String  # structured data types not yet supported
     compressor::C
     fill_value::Union{T, Nothing}
     order::Char
     filters::F  # not yet supported
-    function Metadata{T2, N, C, F}(zarr_format, shape, chunks, dtype, compressor,fill_value, order, filters) where {T2,N,C,F}
-        #We currently only support version 
-        zarr_format == 2 || throw(ArgumentError("Zarr.jl currently only support v2 of the protocol"))
+    chunk_encoding::ChunkEncoding
+    function MetadataV2{T2,N,C,F}(zarr_format, node_type, shape, chunks, dtype, compressor, fill_value, order, filters, chunk_encoding) where {T2,N,C,F}
+        zarr_format == 2 || throw(ArgumentError("MetadataV2 only functions if zarr_format == 2"))
         #Do some sanity checks to make sure we have a sane array
         any(<(0), shape) && throw(ArgumentError("Size must be positive"))
         any(<(1), chunks) && throw(ArgumentError("Chunk size must be >= 1 along each dimension"))
         order === 'C' || throw(ArgumentError("Currently only 'C' storage order is supported"))
-        new{T2, N, C, F}(zarr_format, Base.RefValue{NTuple{N,Int}}(shape), chunks, dtype, compressor,fill_value, order, filters)
+        new{T2,N,C,F}(zarr_format, node_type, Base.RefValue{NTuple{N,Int}}(shape), chunks, dtype, compressor, fill_value, order, filters, chunk_encoding)
     end
 end
+zarr_format(::MetadataV2) = ZarrFormat(Val(2))
+
+"""Metadata for Zarr version 3 arrays"""
+struct MetadataV3{T,N,C,F} <: AbstractMetadata{T,N,C,F}
+    zarr_format::Int
+    node_type::String
+    shape::Base.RefValue{NTuple{N, Int}}
+    chunks::NTuple{N, Int}
+    dtype::String  # data_type in v3
+    compressor::C
```

Comment:
If we are going to implement V3 metadata properly, we could probably remove this and filters and replace with `Codecs`.